### PR TITLE
DuplicateTokenEx

### PR DESCRIPTION
**Pull request by zabulus**

### CODE REVIEW

File: src/AdvApi32.Desktop/AdvApi32+TokenType.cs
Comment (vbfox) [inline src/AdvApi32.Desktop/AdvApi32+TokenType.cs]: This enumeration is named `TOKEN_TYPE` on MSDN : https://msdn.microsoft.com/en-us/library/windows/desktop/aa379633(v=vs.85).aspx

File: src/AdvApi32.Desktop/AdvApi32.cs
Comment (vbfox) [inline src/AdvApi32.Desktop/AdvApi32.cs]: I don't remember the details of discussions about access rights @AArnott should we create a specific `ACCESS_MASK` this time or wait for a decision on unification (#271)

(Knowing that theThe access rights are already present as uint constants in `AdvApi32+TokenAccessRights.cs`)

File: src/AdvApi32.Desktop/AdvApi32.cs
Comment (vbfox) [inline src/AdvApi32.Desktop/AdvApi32.cs]: We have the structure, so it should be of type `SECURITY_ATTRIBUTES*`.

**Review by vbfox — CHANGES_REQUESTED** [review]

File: src/AdvApi32.Desktop/AdvApi32.cs
Comment (AArnott) [reply src/AdvApi32.Desktop/AdvApi32.cs]: Yes, with a `[Friendly(In|Optional)]` attribute on it.

File: src/AdvApi32.Desktop/AdvApi32.cs
Comment (AArnott) [reply src/AdvApi32.Desktop/AdvApi32.cs]: We should use `ACCESS_MASK` for the time being. It will allow conversion to `uint` when necessary but in the meantime at least captures a bit of the strong-typed nature we're going for.

**Review by AArnott — CHANGES_REQUESTED** [review]

File: src/AdvApi32.Desktop/AdvApi32.cs
Comment (arlm) [reply src/AdvApi32.Desktop/AdvApi32.cs]: There are samples for the specific and special flags, if needed, on the inner enumerations.

**Comment by zabulus** [comment]
Squashed commits, after my ugly merge they lost any meaning.
Addressed review issues:
1. TOKEN_TYPE naming
2. ACCESS_MASK
3. SECURITY_ATTRIBUTES (including attribute)
4. Also return BOOL marshalling

**Review by AArnott — APPROVED** [review]

**Comment by AArnott** [comment]
Thanks for your contribution.

**Pull request merged**
